File: EtiveMor.OpenImoCiiCalculator/EtiveMor.OpenImoCiiCalculator.Core/Services/Impl/ShipCapacityCalculatorService.cs
using EtiveMor.OpenImoCiiCalculator.Core.Models;
using EtiveMor.OpenImoCiiCalculator.Core.Models.Enums;

namespace EtiveMor.OpenImoCiiCalculator.Core.Services.Impl
{


    public class ShipCapacityCalculatorService : IShipCapacityCalculatorService
    {
        /// <summary>
        /// Calculates the ship's capacity according to the MEPC.353(78)guidelines
        /// </summary>
        /// <param name="ship">
        /// The ship to calculate the capacity for
        /// </param>
        /// <returns>
        /// The ship's type capacity according to MEPC.353(78)
        /// </returns>
        public double GetShipCapacity(Ship ship)
        {
            return GetShipCapacity(ship.ShipType, ship.DeadweightTonnage, ship.GrossTonnage);
        }

        /// <summary>
        /// Calculates the ship's capacity according to the MEPC.353(78)guidelines
        /// </summary>
        /// <param name="shipType">
        /// The type of the ship (e.g., BulkCarrier, RoRoCargoShip, CruisePassengerShip, etc.)
        /// </param>
        /// <param name="deadweightTonnage">
        /// The deadweight tonnage (DWT) of the ship, which represents the sum of the weights 
        /// of cargo, fuel, fresh water, ballast water, provisions, passengers, and crew.
        /// </param>
        /// <param name="grossTonnage">
        /// The gross tonnage (GT) of the ship, which is a measure of the ship's overall internal 
        /// volume. GT is not a weight measurement but is used to determine various other 
        /// shipping-related values, such as crew size, safety requirements, and registration 
        /// fees.
        /// </param>
        /// <returns>
        /// The ship's type capacity according to MEPC.353(78)
        /// </returns>
        public double GetShipCapacity(ShipType shipType, double deadweightTonnage, double grossTonnage)
        {
            ValidateTonnageParamsSet(shipType, deadweightTonnage, grossTonnage);


            switch (shipType)
            {
                case ShipType.BulkCarrier:
                    return deadweightTonnage >= 279000 ? 279000 : deadweightTonnage;
                case ShipType.GasCarrier:
                    return deadweightTonnage;
                case ShipType.Tanker:
                    return deadweightTonnage;
                case ShipType.ContainerShip:
                    return deadweightTonnage;
                case ShipType.GeneralCargoShip:
                    return deadweightTonnage;
                case ShipType.RefrigeratedCargoCarrier:
                    return deadweightTonnage;
                case ShipType.CombinationCarrier:
                    return deadweightTonnage;
                case ShipType.LngCarrier:
                    return deadweightTonnage < 65000 ? 65000 : deadweightTonnage;
                case ShipType.RoRoCargoShipVehicleCarrier: 
                    return deadweightTonnage >= 57700 ? 57700 : grossTonnage;
                case ShipType.RoRoCargoShip:
                    return grossTonnage;
                case ShipType.RoRoPassengerShip:
                    return grossTonnage;
                case ShipType.RoRoPassengerShip_HighSpeedSOLAS:
                    return grossTonnage;
                case ShipType.RoRoCruisePassengerShip:
                    return grossTonnage;
                default:
                    throw new ArgumentException($"Unsupported {nameof(shipType)}: {shipType}");
            }
        }


        /// <summary>
        /// 
        /// </summary>
        /// <param name="shipType"></param>
        /// <param name="deadweightTonnage">
        /// The ship's deadweight tonnage
        /// Required to be above 0 for ship types:
        ///     - <see cref="ShipType.BulkCarrier"/>
        ///     - <see cref="ShipType.GasCarrier"/>
        ///     - <see cref="ShipType.Tanker"/>
        ///     - <see cref="ShipType.ContainerShip"/>
        ///     - <see cref="ShipType.GeneralCargoShip"/>
        ///     - <see cref="ShipType.RefrigeratedCargoCarrier"/>
        ///     - <see cref="ShipType.CombinationCarrier"/>
        ///     - <see cref="ShipType.LngCarrier"/>
        ///     - <see cref="ShipType.RoRoCargoShip"/>
        /// </param>
        /// <param name="grossTonnage">
        /// The ship's grossTonnage. 
        /// 
        /// Required to be above 0 for ship types: 
        ///     - <see cref="ShipType.RoRoCargoShipVehicleCarrier"/>
        ///     - <see cref="ShipType.RoRoPassengerShip"/>
        ///     - <see cref="ShipType.RoRoCruisePassengerShip"/>
        /// </param>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ArgumentOutOfRangeException">Thrown if the weight value is equal or lower than 0 if it is required to be above 0</exception>
        private void ValidateTonnageParamsSet(ShipType shipType, double deadweightTonnage, double grossTonnage)
        {
            _ = shipType switch
            {
                ShipType.BulkCarrier => ValidateTonnage(deadweightTonnage, nameof(deadweightTonnage), shipType)
                    ? deadweightTonnage
                    : throw new InvalidOperationException(),

                ShipType.GasCarrier => ValidateTonnage(deadweightTonnage, nameof(deadweightTonnage), shipType)
                    ? deadweightTonnage
                    : throw new InvalidOperationException(),

                ShipType.Tanker => ValidateTonnage(deadweightTonnage, nameof(deadweightTonnage), shipType)
                    ? deadweightTonnage
                    : throw new InvalidOperationException(),

                ShipType.ContainerShip => ValidateTonnage(deadweightTonnage, nameof(deadweightTonnage), shipType)
                    ? deadweightTonnage
                    : throw new InvalidOperationException(),

                ShipType.GeneralCargoShip => ValidateTonnage(deadweightTonnage, nameof(deadweightTonnage), shipType)
                    ? deadweightTonnage
                    : throw new InvalidOperationException(),

                ShipType.RefrigeratedCargoCarrier => ValidateTonnage(deadweightTonnage, nameof(deadweightTonnage), shipType)
                    ? deadweightTonnage
                    : throw new InvalidOperationException(),

                ShipType.CombinationCarrier => ValidateTonnage(deadweightTonnage, nameof(deadweightTonnage), shipType)
                    ? deadweightTonnage
                    : throw new InvalidOperationException(),

                ShipType.LngCarrier => ValidateTonnage(deadweightTonnage, nameof(deadweightTonnage), shipType)
                    ? deadweightTonnage
                    : throw new InvalidOperationException(),

                ShipType.RoRoCargoShipVehicleCarrier => ValidateTonnage(grossTonnage, nameof(grossTonnage), shipType)
                    ? grossTonnage
                    : throw new InvalidOperationException(),

                ShipType.RoRoCargoShip => ValidateTonnage(grossTonnage, nameof(grossTonnage), shipType)
                    ? grossTonnage
                    : throw new InvalidOperationException(),

                ShipType.RoRoPassengerShip => ValidateTonnage(grossTonnage, nameof(grossTonnage), shipType)
                    ? grossTonnage
                    : throw new InvalidOperationException(),

                ShipType.RoRoPassengerShip_HighSpeedSOLAS => ValidateTonnage(grossTonnage, nameof(grossTonnage), shipType)
                    ? grossTonnage
                    : throw new InvalidOperationException(),

                ShipType.RoRoCruisePassengerShip => ValidateTonnage(grossTonnage, nameof(grossTonnage), shipType)
                    ? grossTonnage
                    : throw new InvalidOperationException(),

                _ => throw new ArgumentOutOfRangeException(nameof(shipType), shipType, $"Unsupported {nameof(shipType)}: {shipType}")
            };
        }

        /// <summary>
        /// Validates that the tonnage param is greater than 0 
        /// 
        /// </summary>
        /// <param name="tonnage">The tonnage value (accepts either gross tonnage or deadweight)</param>
        /// <param name="tonnageName">The ship's tonnage name (accepts either "gross" or "deadweight"</param>
        /// <param name="shipType">The ship type <seealso cref="ShipType"/></param>
        /// <returns>
        /// true if the tonnage is greater than 0
        /// throws an exception if the tonnage is less than or equal to 0
        /// </returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown if the value is equal or lower than 0</exception>
        private bool ValidateTonnage(double tonnage, string tonnageName, ShipType shipType)
        {
            if (tonnage <= 0)
            {
                throw new ArgumentOutOfRangeException(tonnageName, tonnage, $"{tonnageName} must be greater than 0 if {nameof(shipType)} is set to {shipType} ");
            }
            return true;
        }

    }
}

File: python-impl/open_imo_cii_calculator/services/ship_capacity_calculator_service.py
"""
Service for calculating ship capacity according to MEPC.353(78) guidelines
"""
from typing import Union, Optional
from open_imo_cii_calculator.models.ship_type import ShipType
from open_imo_cii_calculator.models.ship_models.ship import Ship


class ShipCapacityCalculatorService:
    """
    Service for calculating ship capacity according to MEPC.353(78) guidelines
    """
    
    def get_ship_capacity(self, ship_or_type: Union[Ship, ShipType], deadweight_tonnage: Optional[float] = None, gross_tonnage: Optional[float] = None) -> float:
        """
        Calculates the ship's capacity according to the MEPC.353(78) guidelines

        Args:
            - ship_or_type (Union[Ship, ShipType]): The ship object or type of the ship
            - deadweight_tonnage (Optional[float]): The deadweight tonnage of the ship
            - gross_tonnage (Optional[float]): The gross tonnage of the ship

        Returns:
            - float: The calculated ship capacity
        """
        if isinstance(ship_or_type, Ship):
            ship = ship_or_type
            return self.get_ship_capacity(ship.ship_type, ship.deadweight_tonnage, ship.gross_tonnage)
        
        ship_type = ship_or_type
        
        self._validate_tonnage_params_set(ship_type, deadweight_tonnage, gross_tonnage)
        
        if ship_type == ShipType.BULK_CARRIER:
            return 279000 if deadweight_tonnage >= 279000 else deadweight_tonnage
        elif ship_type == ShipType.GAS_CARRIER:
            return deadweight_tonnage
        elif ship_type == ShipType.TANKER:
            return deadweight_tonnage
        elif ship_type == ShipType.CONTAINER_SHIP:
            return deadweight_tonnage
        elif ship_type == ShipType.GENERAL_CARGO_SHIP:
            return deadweight_tonnage
        elif ship_type == ShipType.REFRIGERATED_CARGO_CARRIER:
            return deadweight_tonnage
        elif ship_type == ShipType.COMBINATION_CARRIER:
            return deadweight_tonnage
        elif ship_type == ShipType.LNG_CARRIER:
            return 65000 if deadweight_tonnage < 65000 else deadweight_tonnage
        elif ship_type == ShipType.RORO_CARGO_SHIP_VEHICLE_CARRIER:
            return 57700 if deadweight_tonnage >= 57700 else gross_tonnage
        elif ship_type == ShipType.RORO_CARGO_SHIP:
            return gross_tonnage
        elif ship_type == ShipType.RORO_PASSENGER_SHIP:
            return gross_tonnage
        elif ship_type == ShipType.RORO_PASSENGER_SHIP_HIGH_SPEED_SOLAS:
            return gross_tonnage
        elif ship_type == ShipType.CRUISE_PASSENGER_SHIP:
            return gross_tonnage
        else:
            raise ValueError(f"Unsupported ship type: {ship_type}")
            
    def _validate_tonnage_params_set(self, ship_type: ShipType, deadweight_tonnage: Optional[float], gross_tonnage: Optional[float]):
        """
        Validates that the required tonnage parameters are set for the given ship type
        
        Args:
            - ship_type (ShipType): The type of ship
            - deadweight_tonnage (Optional[float]): The ship's deadweight tonnage
            - gross_tonnage (Optional[float]): The ship's gross tonnage
            
        Raises:
            - ValueError: If the required tonnage parameters are not set or invalid
        """
        # Ship types that require deadweight tonnage
        deadweight_ship_types = [
            ShipType.BULK_CARRIER,
            ShipType.GAS_CARRIER,
            ShipType.TANKER,
            ShipType.CONTAINER_SHIP,
            ShipType.GENERAL_CARGO_SHIP,
            ShipType.REFRIGERATED_CARGO_CARRIER,
            ShipType.COMBINATION_CARRIER,
            ShipType.LNG_CARRIER
        ]
        
        # Ship types that require gross tonnage
        gross_tonnage_ship_types = [
            ShipType.RORO_CARGO_SHIP_VEHICLE_CARRIER,
            ShipType.RORO_CARGO_SHIP,
            ShipType.RORO_PASSENGER_SHIP,
            ShipType.RORO_PASSENGER_SHIP_HIGH_SPEED_SOLAS,
            ShipType.CRUISE_PASSENGER_SHIP
        ]
        
        if ship_type in deadweight_ship_types:
            self._validate_tonnage(deadweight_tonnage, "deadweight_tonnage", ship_type)
            
        if ship_type in gross_tonnage_ship_types:
            self._validate_tonnage(gross_tonnage, "gross_tonnage", ship_type)
            
    def _validate_tonnage(self, tonnage: Optional[float], tonnage_name: str, ship_type: ShipType):
        """
        Validates that the tonnage value is greater than 0
        
        Args:
            - tonnage (Optional[float]): The tonnage value to validate
            - tonnage_name (str): The name of the tonnage parameter
            - ship_type (ShipType): The type of ship
            
        Raises:
            - ValueError: If the tonnage value is less than or equal to 0
        """
        if tonnage is None or tonnage <= 0:
            raise ValueError(f"{tonnage_name} must be greater than 0 if ship_type is set to {ship_type}")
        
        return True